test/studies/adventOfCode/2022/day07/bradc/day07-min.chpl

```diff
@@ -0,0 +1,124 @@
+// A potentially overwrought solution that stored the entire state of
+// the file system in hopes of using it for part 2.  No such luck...
+
+
+// New features used here, not seen on the blog yet:
+
+// classes
+// - default initializers
+// - secondary methods
+// select statements
+// ref intents
+// Recursion (not particularly Chapel-specific
+// assert()
+
+use IO, List;
+
+// A class representing a directory and its contents
+//
+class Dir {
+  var filesizes: list(int);      // list of filenames
+  var subdirs: list(owned Dir);  // list of subdirs
+}
+
+
+// Kick things off by creating a root directory.
+//
+var RootDir = new Dir();
+
+
+// Then read in the first line, making sure it's as expected.
+//
+var line: string;
+
+readLine(line);
+assert(line == "$ cd /\n");
+
+
+// Then read in the rest of the disk's state.
+//
+RootDir.readContents();
+
+
+// Now, compute the sum of all directory sizes less than 'dirThreshold'
+config param dirThreshold = 100_000;
+var sumOfSmall = 0;
+
+RootDir.findSize(dirThreshold, sumOfSmall);
+writeln(sumOfSmall);
+
+//
+// Routine for reading in a directory's contents via terminal session commands
+//
+proc Dir.readContents() {
+  var line: string;
+
+  while readLine(line) {
+    line = line.strip();  // remove the newline
+
+    select line[..3] {  // categorize the command by the first 4 characters
+
+      when "$ ls" {
+        // no need to do anything, we'll just take care of the next lines
+      }
+
+      when "dir " {
+        //
+        // line:  dir <dirname>
+        // token: 0   1
+        // 
+
+        /* borrow idea from Daniel and do nothing here, moving work to
+           'cd' instead */
+      }
+
+      when "$ cd" {
+        //
+        // line:  $ cd <dirname>
+        // token: 0 1  2
+        //
+        const token = line.split(" ");
+        ref dirname = token[2];
+
+        if dirname == ".." {
+          return;  // pop us back up a level
+        } else {
+          var newdir = new Dir();
+          newdir.readContents();
+          subdirs.append(newdir);
+        }
+      }
+
+      otherwise {  // This is a file
+        //
+        // line:   12345 <filename>
+        // token:  0     1
+        //
+        const token = line.split(" "),
+              size = token[0]:int;
+        ref filename = token[1];
+   
+        filesizes.append(size);
+      }
+    }
+  }
+}
+
+
+//
+// Routine for adding up directory sizes less than threshold
+//
+proc Dir.findSize(threshold, ref sum): int {
```

Comment:
@DanilaFe Sorry, this was the closest I could find to leaving a comment on the blog entry, and I don't see that you're active on gitter.  The blog says

> I could’ve omitted the argument to dirSizes — notice from the method’s signature that I provide a default value for parentSize.

```chapel
iter dirSizes(ref parentSize = 0): int {
```

but that's not true for `ref` args -- see #19116.  

Also see my comment in this pr yesterday about `"at most" the threshold size`.

Reply:
Thanks for pinging me, I'd better get on Gitter! 

I was not aware of the issue you link, but I do know that my code compiles, and works as I'd expect -- including when I omit the `parentSize` parameter.

Reply:
(And yet I also get a clean compile from your code in the blog with the omitted arg.  Modifying my code to make the iter recursive, and invoke it only in the same reduction / `[]`-expression syntax still hits the same error.)

Ah, I was able to make my code compile by putting the iter in a `class` instead of a `record`, using a regular iteration as above.  And in a class it still compiles as a `proc` instead of an `iter`.

Reply:
I opened #21213 for the inconsistent behavior here.

Reply:
Thank you for doing that. It's definitely curious that the behavior differs between classes and records in this situation.